platform/darwin/src/NSComparisonPredicate+MGLAdditions.mm

```diff
@@ -361,9 +142,52 @@ - (id)mgl_jsonExpressionObject {
                         format:@"NSPredicateOperatorType:%lu is not supported.", (unsigned long)self.predicateOperatorType];
     }
     if (op) {
-        return @[op, self.leftExpression.mgl_jsonExpressionObject, self.rightExpression.mgl_jsonExpressionObject];
+        id leftJSONExpressionObject = self.leftExpression.mgl_jsonExpressionObject;
+        id rightJSONExpressionObject = self.rightExpression.mgl_jsonExpressionObject;
+        
+        switch (self.predicateOperatorType) {
+            case NSLessThanPredicateOperatorType:
+            case NSLessThanOrEqualToPredicateOperatorType:
+            case NSGreaterThanPredicateOperatorType:
+            case NSGreaterThanOrEqualToPredicateOperatorType: {
+                leftJSONExpressionObject = [self mgl_typedJSONWithExpressionObject:self.leftExpression fallbackExpression:self.rightExpression];
+                rightJSONExpressionObject = [self mgl_typedJSONWithExpressionObject:self.rightExpression fallbackExpression:self.leftExpression];
+                break;
+            }
+ 
+            default:
+                break;
+        }
+        
+        return @[op, leftJSONExpressionObject, rightJSONExpressionObject];
     }
     return nil;
 }
 
+/**
+ Infers the expressionObject type if it can not then tries to infer the type using fallbackExpression.
+ ExpressionFilters of type >, >=, <, <= requires that each compared value provides its type.
+ */
+- (id)mgl_typedJSONWithExpressionObject:(NSExpression *)expressionObject fallbackExpression:(NSExpression *)fallbackExpression {
```

Comment:
That method is generic. I was passing both sides of the comparison expression. So calling self didn't make sense.

I removed this code since it's purpose to aid the developers in the migration is not working as intended and is causing more problems per mi comment https://github.com/mapbox/mapbox-gl-native/pull/11587#discussion_r181504893.

Reply:
For future reference, it would’ve been possible to call, e.g., `[leftExpression mgl_typeCheckedExpressionMatchingExpression:rightExpression]`, in which case `self` inside the method would’ve been equivalent to `leftExpression` at the call site.